### PR TITLE
Use format_version=6 by default

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: It's in production for a large storage service, and it was initially released 6 months ago (8.6.0). IMHO that's enough room for "easy downgrade" to most any user's previously integrated version, even if they only update a few times a year.

Test Plan: tests updated, including format capatibility test

table_test: ApproximateOffsetOfCompressed is affected because adding index block to metaindex adds about 13 bytes
to SST files in format_version 6. This test has historically been problematic and one reason is that, apparently, not only
could it pass/fail depending on snappy compression version, but also how long your host name is, because of db_host_id.
I've cleared that out for the test, which takes care of format_version=6 and hopefully improves long-term reliability.

Suggested follow-up: FinishImpl in table_test.cc takes a table_options that is ignored in some cases and might not match
the ioptions.table_factory configuration unless the caller is very careful. This should be cleaned up somehow.

### CODE REVIEW

File: unreleased_history/behavior_changes/format6.md [hunk 1]
```diff
@@ -0,0 +1 @@
+format\_version=6 is the new default setting in BlockBasedTableOptions, for more robust data integrity checking. DBs and SST files written with this setting cannot be read by RocksDB versions before 8.6.0.
```
Comment (anand1976) [inline unreleased_history/behavior_changes/format6.md:1]: Maybe worth mentioning explicitly that it doesn't apply to plain table format.

**Review by anand1976 — APPROVED** [review]
LGTM

**Comment by siying** [comment]
@pdillinger We are explicitly setting format_version to 5 (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala#L978), I wonder which version can still provide forward compatibility. Half a year forward compatibility might be enough for Meta but far from enough for us, so I wonder how much it is if we pin the format version. Since format compatibility test doesn't have such a test, I can't tell from it.

**Comment by ajkr** [comment]
The release notes say format_version 5 was introduced in version 6.6 (Nov 2019), so you should be able to downgrade much more than you'd really want to, considering all the bug fixes that were made over the past 4+ years.

**Comment by siying** [comment]
@ajkr thanks for your answer. In my understanding, table format version isn't the only place that can introduce forward compatibility issues. For example, manifest format change or WAL format change can also cause some. I guess these didn't change recently?

**Comment by ajkr** [comment]
Good question. There are not a lot of cases where `check_format_compatible.sh` changed the base version for forward compatibility checking. The last one before 8.6 for `format_version == 6` was 6.27 for `kXXH3`. I guess it's still possible that there's something else that would prevent downgrading, for example, from 9.0 to 7.0 while using `format_version == 5`. Maybe you can adapt the compatibility script to use your options and see how far it gets?

**Pull request closed without merging**